### PR TITLE
Add ServiceUrlProvider and add method forceCloseConnection in PulsarC…

### PR DESCRIPTION
**Pull request by codelipenghui**
Support build Pulsar client with serviceUrlProvider method.

### Motivation

With serviceUrlProvider we can store the pulsar service url in zookeeper or any other config service.
And we can watch the service url change event then control the pulsar client, such as change pulsar client serviceUrl, force close client connection or re-connect with new service url.

### Modifications

Add ServiceUrlProvider interface.
Add forceCloseConnection method in PulsarClient.

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java [hunk 1]
```diff
@@ -706,6 +706,58 @@ public void shutdown() throws PulsarClientException {
         }
     }
 
+    @Override
+    public void forceCloseConnection() {
+        if (this.producers != null) {
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java]: There are few problems with this way of closing the connections. 

`channelInactive()` is supposed to be called by Netty when the connection gets closed. With the current approache the original TCP connection will stay open and the broker will still have registered the producer/consumer as connected. 

The right approach here should be to close all the connections in the `ConnectionPool`. That will be triggering the reconnections for all producers/consumers.

Comment (codelipenghui) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java]: @merlimat I'm already fix it.

**Comment by sijie** [comment]
run java8 tests

**Comment by sijie** [comment]
@merlimat can you review this PR?

**Comment by sijie** [comment]
retest this please

**Comment by sijie** [comment]
run java8 tests

**Comment by jiazhai** [comment]
over all lgtm, it would be better provide more comments and examples as @sijie suggested.

**Comment by codelipenghui** [comment]
Thanks @sijie.

We will implement a service url provider base on Zookeeper. If this url provider have a good practice in zhaopin.com. we pleasure to contribute a Pulsar Client extend.

**Comment by sijie** [comment]
@codelipenghui yeah, thank you so much for your contribution! look forward to your contribution about zookeeper based service url provider :)

**Pull request merged**